l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,478 @@
+
+        Node storage node = nodes[_nodeOwner];
+        ensureAttesterSnapshot(node);
+        node.attesterLatest.weight = _weight;
+
+        emit NodeAttesterWeightChanged(_nodeOwner, _weight);
+    }
+
+    /// @notice Changes the validator's public key and proof-of-possession (PoP) in the registry.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner whose validator key and PoP will be changed.
+    /// @param _pubKey The new BLS12-381 public key to assign to the node's validator.
+    /// @param _pop The new proof-of-possession (PoP) to assign to the node's validator.
+    function changeValidatorKey(
+        address _nodeOwner,
+        BLS12_381PublicKey calldata _pubKey,
+        BLS12_381Signature calldata _pop
+    ) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        verifyInputBLS12_381PublicKey(_pubKey);
+        verifyInputBLS12_381Signature(_pop);
+        verifyNodeOwnerExists(_nodeOwner);
+
+        Node storage node = nodes[_nodeOwner];
+        ensureValidatorSnapshot(node);
+        node.validatorLatest.pubKey = _pubKey;
+        node.validatorLatest.pop = _pop;
+
+        emit NodeValidatorPubKeyChanged(_nodeOwner, _pubKey);
+        emit NodeValidatorPoPChanged(_nodeOwner, _pop);
+    }
+
+    /// @notice Changes the attester's public key of a node in the registry.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner whose attester public key will be changed.
+    /// @param _pubKey The new ECDSA public key to assign to the node's attester.
+    function changeAttesterPubKey(
+        address _nodeOwner,
+        Secp256k1PublicKey calldata _pubKey
+    ) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        verifyInputSecp256k1PublicKey(_pubKey);
+        verifyNodeOwnerExists(_nodeOwner);
+
+        Node storage node = nodes[_nodeOwner];
+        ensureAttesterSnapshot(node);
+        node.attesterLatest.pubKey = _pubKey;
+
+        emit NodeAttesterPubKeyChanged(_nodeOwner, _pubKey);
+    }
+
+    /// @notice Rotates the validator committee list based on active nodes in the registry.
+    /// @dev Only callable by the contract owner.
+    function commitValidators() external onlyOwner {
+        validatorsCommit++;
+
+        uint256 len = pendingRemovals.length;
+        for (uint256 i = 0; i < len; ++i) {
+            PendingNodeRemoval storage entry = pendingRemovals[i];
+            entry.pendingValidatorsCommit = false; // Validator is considered as removed.
+            if (!entry.pendingAttestersCommit) {
+                finalizeNodeRemoval(entry.nodeOwner);
+            }
+        }
+
+        emit ValidatorsCommitted();
+    }
+
+    /// @notice Rotates the attester committee list based on active nodes in the registry.
```

Comment:
Sorry if I'm lacking the context here, but where is this rotation taking place? All I see is the removal of pending nodes and the increment of the commit number. I don't see the `.active` flag being read anywhere in the contract, nor the various snapshots for that matter. Shouldn't it create the next cohort of `CommitteeAttester` records? Initially I thought it might create a new snapshot, but that only happens when something is updated.

Reply:
I agree that it’s quite confusing. 
I just [added](https://github.com/matter-labs/era-contracts/pull/555/commits/89b4ccdd038331730f6ef4a488364090778b18b4) the docs, so it could help.

The source discussion of this change is here: https://github.com/matter-labs/era-contracts/pull/555#discussion_r1705431453

Reply:
Great, thanks for the docs that made is much easier to understand. So readers will have to interact with the storage itself, there is not read-only method to call to get the list of validators?

Reply:
If you refer to “storage” as the contract state data structures (and not the VM implementation KV database), then yes, readers would need to interact with these, as previously, just with some extra complexity.
Having read-only methods that would expose the committee lists would require to serialize on-chain into blobs (to storage or events), and so introduce additional gas costs.

Reply:
I'm not sure I understand exactly what you mean. Probably just a terminology thing, please bear with me.

Basically I wanted to understand how these offline readers can get the information they need to select the active validators. In my previous experience with a similar Solidity contract managing validators, the contract had methods like `function getActiveValidators() external view returns (WeightedValidator[] memory)` that I could call from Rust to get the list, and it encapsulated all the logic required for filtering. For example in this case it would look at the `commitValidators`, etc, and the caller wouldn't have to understand the `snapshot` vs `latest` distinction. I called these on the local VM with an `eth_call`, as you do in your extractor, so they wouldn't cost any real gas, but of course they still cost a VM execution.

Since we I don't see such methods, I thought that the offline client (e.g. your extractor) will have to interact directly with the storage layout of the contract. I haven't seen such an example, but didn't imagine it would be easy - to my mind that uses `eth_getStorageAt` with the U256 address of where Solidity stored the map entries, returning U256 words for data. It sounds complicated to use these directly. 

I didn't think about interacting with the underlying KV database directly, which I imagine has even more indirection. 

I don't understand the extra serialisation into blobs you mentioned. Like I say I thought we would just let the contract read its own storage and return the DTOs it defined in `IContractRegistry.sol`.

Reply:
Following our discussion on the daily, now I understand what you mean by interacting with the storage: for example [here](https://github.com/matter-labs/zksync-era/blob/e9b7c364e08bfbc5c977f60e88d40d9f881dc684/core/node/consensus/src/storage/vm_reader.rs#L202) you're interacting with a field by calling `my_contract.function("my_field_name")` and then using `eth_call` to invoke it with the node owner as input; at least that's what it looks like, it might be an actual function in that instance. But [here](https://github.com/matter-labs/zksync-era/blob/e9b7c364e08bfbc5c977f60e88d40d9f881dc684/core/node/consensus/src/storage/vm_reader.rs#L153) it doesn't even have a parameter, so maybe that's the example. I didn't know this was possible. 

And based on [this discussion](https://github.com/matter-labs/zksync-era/pull/2416/files/e9b7c364e08bfbc5c977f60e88d40d9f881dc684#r1671949109) I understand that executing `eth_call` has some baggage in that the sandbox will [open a new DB connection](https://github.com/matter-labs/zksync-era/pull/2416/files/e9b7c364e08bfbc5c977f60e88d40d9f881dc684#r1671942042) every time, and that's one of the reasons direct storage _slot_ access was brought up as an alternative.

I just wanted to add that if there was an accessor method I described above to encapsulate the listing of active validators/attesters, then we wouldn't need [multiple reads](https://github.com/matter-labs/zksync-era/blob/e9b7c364e08bfbc5c977f60e88d40d9f881dc684/core/node/consensus/src/storage/vm_reader.rs#L89-L94) because that one `eth_call` to the "smart" method could do everything we need.

Reply:
> I just wanted to add that if there was an accessor method I described above to encapsulate the listing of active validators/attesters, then we wouldn't need [multiple reads](https://github.com/matter-labs/zksync-era/blob/e9b7c364e08bfbc5c977f60e88d40d9f881dc684/core/node/consensus/src/storage/vm_reader.rs#L89-L94) because that one eth_call to the "smart" method could do everything we need.

yes @aakoshh this is a good point - although we could also read multiple values in one eth-call, it's more clean to have an accessor method

Reply:
IMO it also makes sense from the contract design prospective - you show the indended usage right away in the interface

Reply:
A good point indeed.

In the previous design it would be arguable, since the contract simply stored `[]CommitteeValidator` to be read. In that case, adding on-chain optimization for readers, instead of optimizing the VM implementation, might be a bad tradeoff. 

However, I completely overlooked that returning a `memory` array from a function, instead reading it from storage, would not require N read transaction. So that small optimization would have been worthwhile anyway.

With the new design, not encapsulating the read logic is pretty bad, and leaks part of the contract implementation off-chain. 

I added 2 getter functions here: https://github.com/matter-labs/era-contracts/pull/555/commits/262adb6e342c979b299205a55e09bec281d13b89

The main caveat is that it defines the committee selection logic (all active and non-removed nodes) on-chain, which makes it less flexible for the next versions.